### ISSUE TITLE
Manual specification of step order

### ISSUE THREAD

**Issue by wgantt** [allenai/tango#418]
### 🚀 The feature, motivation and pitch

As far as I can tell (and I may be wrong about this), the only thing that dictates parent-child relations in the step graph is direct chaining of outputs to inputs via "ref". But it would be nice to have a way to _manually_ enforce that a step (S2) executes after another (S1), even if S1 is not an ancestor of S2 in the graph determined by "ref" relations. Is there a way to do this?

### Alternatives

A naïve solution to this would be to (e.g.) have the `run` method of S2 take the outputs of S1 as an argument and simply not use that argument. But this seems inelegant.

### Additional context

My personal use case motivating this is a set of LM prompting experiments in which one of the steps in the experiment simply prints a set of randomly sampled, dynamically constructed prompts. I would like to see what these prompts look like before model inference runs (as a separate step) in order to be confident the inputs are correct, but currently the prompts print only after the inference step. And the printing step and the model inference step are siblings in the step graph.

I am using the latest tango release (v0.14.0).

**Comment by epwalsh** [comment]
Hey @wgantt, this does seem like a good feature, and you're right the currently the only way to enforce a parent-child relationship is through inputs/outputs. I'm thinking we could add another `Step` parameter like `step_requires: List[Union[Step, str]]` that would take a list of steps / step names / step unique IDs that must be run before the current step.

So the config in your example might look something like this:

```jsonnet
{
  steps: {
    print_prompts: { type: "print_prompts" },
    eval_prompts: { type: "eval_prompts", step_requires: [ "print_prompts" ] },
  },
}
```

**Comment by wgantt** [comment]
Yep! At least from a user's perspective, this seems like the most sensible way to do it.

**Comment by epwalsh** [comment]
I made one change to my proposal in #419. This example would be:

```jsonnet
{
  steps: {
    print_prompts: { type: "print_prompts" },
    eval_prompts: { type: "eval_prompts", step_requires: [ { type: "ref", ref: "print_prompts" } ] },
  },
}
```

That is, dependencies in `step_requires` need to be expressed in the "ref step" format.

**Comment by wgantt** [comment]
This looks great — thanks so much for the quick turnaround @epwalsh!
